_posts/2023-09-06-load-balancer.md

```diff
@@ -0,0 +1,906 @@
+
+```scala
+package com.ghurtchu.loadbalancer.services
+
+import cats.effect.IO
+import cats.effect.unsafe.implicits.global
+import com.ghurtchu.loadbalancer.domain.Urls
+import com.ghurtchu.loadbalancer.domain.Urls._
+import com.ghurtchu.loadbalancer.domain.Backends._
+import com.ghurtchu.loadbalancer.services.SendAndExpect._
+import munit.FunSuite
+
+class UpdateRefUrlsAndGetTest extends FunSuite {
+
+  val updateBackendsAndGet = UpdateBackendsAndGet.impl
+  val localhost            = Url("localhost:8083")
+
+  test("Alive") {
+    val status = Status.Alive
+    val urls   = Urls(Vector("localhost:8081", "localhost:8082"))
+
+    (for {
+      ref     <- IO.ref(urls)
+      updated <- updateBackendsAndGet(Backends(ref), localhost, status)
+    } yield updated.values == (urls.values :+ localhost))
+      .unsafeRunSync()
+  }
+
+  test("Dead") {
+    val status = Status.Dead
+    val urls   = Urls(Vector("localhost:8081", "localhost:8082", localhost))
+
+    (for {
+      ref     <- IO.ref(urls)
+      updated <- updateBackendsAndGet(Backends(ref), localhost, status)
+    } yield updated.values == Vector("localhost:8081", "localhost:8082"))
+      .unsafeRunSync()
+  }
+
+}
+```
+
+With these services available we have everything we need to implement two main bosses:
+- load balancer
+- backend health checks
+
+Let's start with the simpler one - `HealthCheckBackends`.
+
+This service will use definitions from `domain` and `services` to check the availability of backends, so effectively it will:
+- apply Round Robin to `HealthChecks`
+- parse uri string to `Uri`
+- ping the `Uri`
+- sleep for the predetermined amount of time and do the same again (functional `while (true)` loop if I can say so)
+
+Let's express this idea as code:
+```scala
+package com.ghurtchu.loadbalancer.services
+
+import cats.effect.IO
+import com.ghurtchu.loadbalancer.domain.Config.HealthCheckInterval
+import com.ghurtchu.loadbalancer.domain.Backends.{Backends, HealthChecks}
+import com.ghurtchu.loadbalancer.services.SendAndExpect.Status
+import com.ghurtchu.loadbalancer.services.RoundRobin.HealthChecksRoundRobin
+
+import scala.concurrent.duration.DurationLong
+
+object HealthCheckBackends {
+
+  def periodically(
+    healthChecks: HealthChecks,
+    backends: Backends,
+    parseUri: ParseUri,
+    updateBackendsAndGet: UpdateBackendsAndGet,
+    healthChecksRoundRobin: HealthChecksRoundRobin,
+    sendAndExpectStatus: SendAndExpect[Status],
+    healthCheckInterval: HealthCheckInterval
+  ): IO[Unit] =
+    (for {
+      currentUrl <- healthChecksRoundRobin(healthChecks)
+      uri        <- IO.fromEither(parseUri(currentUrl.value))
+      status     <- sendAndExpectStatus(uri)
+      _          <- updateBackendsAndGet(backends, currentUrl, status)
+    } yield ())
+      .flatMap(_ => IO.sleep(healthCheckInterval.value.seconds))
+      .foreverM
+}
+```
+
+As you can see `HealthCheckBackends` is just a piece of logic which uses the services which we defined above.
```

Comment:
We need to explain this piece of logic in more detail. So far we've built a bunch of components that we need to tie together. At this point it's not clear where each component fits, unless the reader tries to dig into the types themselves.